src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Custom/Serializer/ActivitySerializer.cs

```diff
@@ -0,0 +1,428 @@
+        {
+            cursor = Writer.WriteStringWithTag(ref buffer, cursor, FieldNumberConstants.Span_trace_state, activity.TraceStateString);
+        }
+
+        cursor = Writer.WriteEnumWithTag(ref buffer, cursor, FieldNumberConstants.Span_kind, (int)activity.Kind + 1);
+        cursor = Writer.WriteFixed64WithTag(ref buffer, cursor, FieldNumberConstants.Span_start_time_unix_nano, (ulong)activity.StartTimeUtc.ToUnixTimeNanoseconds());
+        cursor = Writer.WriteFixed64WithTag(ref buffer, cursor, FieldNumberConstants.Span_end_time_unix_nano, (ulong)(activity.StartTimeUtc.ToUnixTimeNanoseconds() + activity.Duration.ToNanoseconds()));
+        cursor = Writer.WriteTagAndLengthPrefix(ref buffer, cursor, 16, FieldNumberConstants.Span_trace_id, WireType.LEN);
+        cursor = SerializeTraceId(ref buffer, cursor, activity.TraceId);
+        cursor = Writer.WriteTagAndLengthPrefix(ref buffer, cursor, 8, FieldNumberConstants.Span_span_id, WireType.LEN);
+        cursor = SerializeSpanId(ref buffer, cursor, activity.SpanId);
+        if (activity.ParentSpanId != default)
+        {
+            cursor = Writer.WriteTagAndLengthPrefix(ref buffer, cursor, 8, FieldNumberConstants.Span_parent_span_id, WireType.LEN);
+            cursor = SerializeSpanId(ref buffer, cursor, activity.ParentSpanId);
+        }
+
+        cursor = this.SerializeActivityTags(ref buffer, cursor, activity, out var statusCode, out var statusMessage);
+        cursor = SerializeActivityStatus(ref buffer, cursor, activity, statusCode, statusMessage);
+        cursor = this.SerializeActivityEvents(ref buffer, cursor, activity);
+        cursor = this.SerializeActivityLinks(ref buffer, cursor, activity);
+        cursor = SerializeTraceFlags(ref buffer, cursor, activity.ActivityTraceFlags, activity.HasRemoteParent, FieldNumberConstants.Span_flags);
+        return cursor;
+    }
+
+    private int SerializeActivityTags(ref byte[] buffer, int cursor, Activity activity, out StatusCode? statusCode, out string? statusMessage)
+    {
+        statusCode = null;
+        statusMessage = null;
+        int maxAttributeCount = this.sdkLimitOptions.SpanAttributeCountLimit ?? int.MaxValue;
+        int maxAttributeValueLength = this.sdkLimitOptions.AttributeValueLengthLimit ?? int.MaxValue;
+        int attributeCount = 0;
+        int droppedAttributeCount = 0;
+        foreach (ref readonly var tag in activity.EnumerateTagObjects())
+        {
+            switch (tag.Key)
+            {
+                case SpanAttributeConstants.StatusCodeKey:
+                    statusCode = StatusHelper.GetStatusCodeForTagValue(tag.Value as string);
+                    continue;
+                case SpanAttributeConstants.StatusDescriptionKey:
+                    statusMessage = tag.Value as string;
+                    continue;
+            }
+
+            if (attributeCount < maxAttributeCount)
+            {
+                /*
+                    // Alternate approach
```

Comment:
@CodeBlanch - Could you take a look at this to see if it matches your expectation? I have not done the complete implementation yet, but can change if the direction looks good to you.